File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    isSame(p,q){
        if(p==q) return true
        if(!p || !q) return false
        if(p.val==q.val){
            return this.isSame(p.left,q.left) && this.isSame(p.right,q.right)
        }else{
            return false
        }
    }
    /**
     * @param {TreeNode} root
     * @param {TreeNode} subRoot
     * @return {boolean}
     */
    isSubtree(root, subRoot) {
        if(root==subRoot){
            return true
        }
        if(!root || !subRoot) return false
        return this.isSame(root,subRoot) || this.isSubtree(root.left,subRoot) || this.isSubtree(root.right,subRoot)
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
      return self.dfs(root)[1]
    def dfs(self,cur):
      if not cur:
        return 0,True
      left,right=self.dfs(cur.left),self.dfs(cur.right)
      if not left[1] or not right[1] or abs(right[0]-left[0])>1:
        return 0,False
      
      return 1+max(left[0],right[0]),True